### PR TITLE
feat: implement tracker

### PR DESCRIPTION
**Pull request by iGxnon**
Please briefly answer these questions:

* what problem are you trying to solve? (or if there's no problem, what's the motivation for this change?)

   Add tracker.

* what changes does this pull request make?

   Add `tracker.rs`.

* are there any non-obvious implications of these changes? (does it break compatibility with previous versions, etc)

    No

### CODE REVIEW

File: crates/curp/src/tracker.rs [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+/// A one-direction bit vector queue
+/// It use a ring buffer `VecDeque<usize>` to store bits
+///
+/// Memory Layout:
+///
+/// `010000000101_110111011111_000001000000000`
+///  ^        ^                ^    ^
+///  |        |________________|____|
+///  |________|         len    |____|
+///     head                    tail
```
Comment (themanforfree) [inline crates/curp/src/tracker.rs:24]: tail is a part of len?

File: crates/curp/src/tracker.rs [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+    fn front(&self) -> Option<bool> {
+        let front = self.store.front()?;
+        Some(front & (1 << (USIZE_BITS - 1 - self.head)) != 0)
+    }
+
+    /// Peek the front usize in store and take a function to check it,
+    /// Used to batch accelerate check.
+    fn front_by(&self, f: impl Fn(usize) -> bool) -> Option<bool> {
+        self.store.front().copied().map(f)
+    }
```
Comment (themanforfree) [inline crates/curp/src/tracker.rs:119]: I think it would be better to name it `map_front`

File: crates/curp/src/tracker.rs (hunk 1 above)
Comment (iGxnon) [reply crates/curp/src/tracker.rs:24]: Tail indicates how many bits are used in the last usize.

File: crates/curp/src/tracker.rs (hunk 2 above)
Comment (iGxnon) [reply crates/curp/src/tracker.rs]: done

File: crates/curp/src/tracker.rs [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+            tail: 0,
+        }
+    }
+
+    /// Get the length
+    fn len(&self) -> usize {
+        if self.store.is_empty() {
+            return 0;
+        }
+        (self.store.len() - 1) * USIZE_BITS + self.tail - self.head + 1
```
Comment (bsbds) [inline crates/curp/src/tracker.rs:58]: Is `self.tail` always greater than or equal to `self.head + 1`?

File: crates/curp/src/tracker.rs [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+    }
+
+    /// Split this bit vec queue to `at`
+    /// e.g.
+    /// 001100 -> `split_at(2)` -> 1100
+    fn split_at(&mut self, at: usize) {
+        let idx = self.head + at;
+        let index = idx / USIZE_BITS;
+        let slot = idx % USIZE_BITS;
+        if index == self.store.len() - 1 && slot > self.tail {
```
Comment (bsbds) [inline crates/curp/src/tracker.rs:163]: Same as above.

File: crates/curp/src/tracker.rs [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+        }
+        (self.store.len() - 1) * USIZE_BITS + self.tail - self.head + 1
+    }
+
+    /// Get the bit value
+    fn get(&self, idx: usize) -> Option<bool> {
+        let idx = self.head + idx;
+        let index = idx / USIZE_BITS;
+        let slot = idx % USIZE_BITS;
+        if index == self.store.len() - 1 && slot > self.tail {
```
[the remark this answers is not in the document]
Comment (iGxnon) [reply crates/curp/src/tracker.rs:66]: When len == 0, it seems could be covered with
```rust
let bits = self.store.get(index)?; // always return None when len(store) = 0
```

and

```rust
let Some(bits) = self.store.get_mut(index) else {
        // always return when len(store) = 0
        return;
};
```

File: crates/curp/src/tracker.rs [hunk 6]
```diff
@@ -0,0 +1,351 @@
...
+        let bits = self.store.get(index)?;
+        Some(bits & (1 << (USIZE_BITS - 1 - slot)) != 0)
+    }
+
+    /// Set the bit value
+    fn set(&mut self, idx: usize, v: bool) {
+        let idx = self.head + idx;
+        let index = idx / USIZE_BITS;
+        let slot = idx % USIZE_BITS;
+        if index == self.store.len() - 1 && slot > self.tail {
```
[the remark this answers is not in the document]
Comment (iGxnon) [reply crates/curp/src/tracker.rs:78]: Check above.

File: crates/curp/src/tracker.rs (hunk 3 above)
Comment (iGxnon) [reply crates/curp/src/tracker.rs:58]: No, head could equals to tail, when they are both 0, which means that the bitvec is empty.

File: crates/curp/src/tracker.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (iGxnon) [reply crates/curp/src/tracker.rs:68]: Add an empty checking.

**Review by bsbds — APPROVED** [review]

**Review by themanforfree — APPROVED** [review]

**Pull request merged**